File: src/main/java/com/bitwise/dominos/basicingredients/Crust.java
package com.bitwise.dominos.basicingredients;

/**
 * Created by Aniruddha Sinha on 24-06-2016.
 */
public class Crust {
    private String name;
    private double price;

    public Crust(String name, double price) {
        this.name = name;
        this.price = price;
        validateCrustName();
        validatePrice();
    }

    private void validatePrice() {
        if(this.price<0) throw new Crust.InvalidCrustPriceException();
    }

    private void validateCrustName() {
        if(this.name.isEmpty()) throw new Crust.EmptyCrustNameException();
    }

    public String getName() {
        return name;
    }

    public double getPrice() {
        return price;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Crust crust = (Crust) o;

        if (Double.compare(crust.price, price) != 0) return false;
        return name.equals(crust.name);

    }

    @Override
    public int hashCode() {
        int result;
        long temp;
        result = name.hashCode();
        temp = Double.doubleToLongBits(price);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        return result;
    }

    @Override
    public String toString() {
        return "Crust{" +
                "name='" + name + '\'' +
                ", price=" + price +
                '}';
    }

    public class EmptyCrustNameException extends RuntimeException{
    }

    public class InvalidCrustPriceException extends RuntimeException {
    }
}

File: src/main/scala/aniruddha/dominos/basic_ingredients/Crust.scala
package aniruddha.dominos.basic_ingredients

/**
  * Created by Aniruddha Sinha on 26-06-2016.
  */
class Crust(n:String,p:Double) {
  val name=n
  val price=p
  def getName:String=name
  def getPrice:Double=price
  override def toString:String="Crust["+name+","+price+"]"

  def canEqual(other: Any): Boolean = other.isInstanceOf[Crust]

  override def equals(other: Any): Boolean = other match {
    case that: Crust =>
      (that canEqual this) &&
        name == that.name &&
        price == that.price
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(name, price)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}
object Crust{
  def apply(name: String, price: Double): Crust =
    if(name.isEmpty) throw new InvalidCrustNameException
    else if(price<0) throw new InvalidCrustPriceException
    else new Crust(name,price)

  case class InvalidCrustNameException() extends RuntimeException
  case class InvalidCrustPriceException()extends RuntimeException
}